File: example/ast.rs
/// A program is a list of statements
#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(tag = "$kind")]
pub enum Stmt {
    /// if-then-else statement
    IfStmt(Box<IfStmt>),
    /// an expression is also sometimes considered a statement
    ExprStmt(Box<ExprStmt>),
    /// a block is a list of statements
    BlockStmt(Box<BlockStmt>),
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct IfStmt {
    /// if condition
    pub cond: Expr,
    /// then statement
    pub then: Stmt,
    /// else statement, can be empty
    pub otherwise: Option<Stmt>,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct ExprStmt {
    pub expr: Expr,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct BlockStmt {
    pub stmts: Vec<Stmt>,
}

/// An expression is a literal, an identifier, or a binary expression
/// 
/// All I want is a room somewhere,
/// Far away from the cold night air,
/// With one enormous chair,
/// Aow, wouldn't it be loverly?
#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(tag = "$kind")]
pub enum Expr {
    /// numeric literal expression
    NumericLiteral(NumericLiteral),
    /// string literal expression
    StringLiteral(StringLiteral),
    /// boolean literal expression
    BoolLiteral(BoolLiteral),
    /// identifier expression
    Identifier(Identifier),
    /// binary expression
    BinaryExpr(Box<BinaryExpr>),
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct NumericLiteral {
    pub value: i64,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct StringLiteral {
    pub value: String,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct BoolLiteral {
    pub value: bool,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Identifier {
    pub name: String,
}

#[derive(Archive, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct BinaryExpr {
    /// binary operator
    pub op: String,
    /// left operand
    pub left: Expr,
    /// right operand
    pub right: Expr,
}

File: example/ast.ts
/** A program is a list of statements */
export type Stmt = 
    | IfStmt
    | ExprStmt
    | BlockStmt

export interface StmtBase<K extends string> {
    $kind: K;
}

/** if-then-else statement */
export interface IfStmt extends StmtBase<"IfStmt"> {
    /** if condition */
    cond: Expr,
    /** then statement */
    then: Stmt,
    /** else statement, can be empty */
    otherwise?: Stmt,
}

/** an expression is also sometimes considered a statement */
export interface ExprStmt extends StmtBase<"ExprStmt"> {
    expr: Expr,
}

/** a block is a list of statements */
export interface BlockStmt extends StmtBase<"BlockStmt"> {
    stmts: Stmt[],
}

/**
 * An expression is a literal, an identifier, or a binary expression
 * 
 * All I want is a room somewhere,
 * Far away from the cold night air,
 * With one enormous chair,
 * Aow, wouldn't it be loverly?
 */
export type Expr = 
    | NumericLiteral
    | StringLiteral
    | BoolLiteral
    | Identifier
    | BinaryExpr

export interface ExprBase<K extends string> {
    $kind: K;
}

/** numeric literal expression */
export interface NumericLiteral extends ExprBase<"NumericLiteral"> {
    value: number,
}

/** string literal expression */
export interface StringLiteral extends ExprBase<"StringLiteral"> {
    value: string,
}

/** boolean literal expression */
export interface BoolLiteral extends ExprBase<"BoolLiteral"> {
    value: boolean,
}

/** identifier expression */
export interface Identifier extends ExprBase<"Identifier"> {
    name: string,
}

/** binary expression */
export interface BinaryExpr extends ExprBase<"BinaryExpr"> {
    /** binary operator */
    op: string,
    /** left operand */
    left: Expr,
    /** right operand */
    right: Expr,
}
